W5500.device.nut

```diff
@@ -0,0 +1,2538 @@
+    // ***************************************************************************
+    constructor(interruptPin, spi, cs = null, resetPin = null) {
+
+        _interruptPin = interruptPin;
+        _spi = spi;
+        _cs = cs;
+        _resetPin = resetPin;
+        _connections = {};
+        _availableSockets = [];
+
+        // Check the pins
+        if (resetPin) _resetPin.configure(DIGITAL_OUT, 1);
+
+        local imp005 = ("spi0" in hardware);
+        if (_cs != null) {
+            _cs.configure(DIGITAL_OUT, 1);
+        } else if (!imp005) {
+            throw "You must pass in a chip select pin."
+        }
+
+        // Reset the hardware
+        reset();
+
+    }
+
+
+    // ***************************************************************************
+    // reset, note this is blocking for 0.2s
+    // Returns: this
+    // Parameters:
+    //          sw(optional) - boolean if true forces a software reset
+    //          Note: datasheet for W5500 states that software reset is
+    //                unreliable - don't use
+    // **************************************************************************
+    function reset(sw = false) {
+
+        // Stop the interrupt pin from firing
+        _interruptPin.configure(DIGITAL_IN);
+
+        // Clear the connections and sockets
+        _connections = {};
+        _availableSockets = [];
+
+        // Force disconnect/close all ports
+        forceCloseAllSockets();
+
+        // Reset chip to default state, blocks for 0.2s
+        // Note: Datasheet states that W5500 software reset is not reliable, use hardware reset.
+        if (sw || _resetPin == null) {
+            setMode(W5500_SW_RESET);
+            imp.sleep(0.2);
+        } else {
+            _resetPin.write(0);
+            imp.sleep(0.01); // hold at least 500us after assert low
+            _resetPin.write(1);
+        }
+        imp.sleep(0.2); // wait at least 150ms before configuring
+
+        // Reset the default memory allocation
+        _setMemDefaults();
+
+        // server.log("Wiznet chip version: " + getChipVersion());
+
+        return this;
+    }
+
+
+
+    // ***************************************************************************
+    // init(cb)
+    // Returns: nothing
+    // Parameters:
+    //        cb                        - callback function
+    //        remaining(optional)       - for first time Initialise == null
+    //                                  - subseuqent goes to to else
+    //        closes all sockets and fires the callback when complete
+    // ***************************************************************************
+    function init(cb) {
+
+        // Configure interrupts
+        setInterrupt(W5500_CONFLICT_INT_TYPE);
+        clearInterrupt();
+        clearSocketInterrupts();
+        _interruptPin.configure(DIGITAL_IN_PULLUP, handleInterrupt.bindenv(this));
+
+        // Set the default mac address
+        setSourceHWAddr(imp.getmacaddress(), true);
```

Comment:
The Wiznet chip should not use the imp's mac address.  To avoid collisions Hugo suggests using a combo of his mpeg prefix (since he owns these) and the rest should start with F and then be based on the imp's device id.
Hugo's mpeg prefix 00:02:D7  
Wiznet's prefix is 00:08:DC